[WebAuthn] Overrelease in error for multi-origin credentials

pascoej opened the pull request:
#### be2916dac989bfb95ddfe2acb75c9631f1d00959
<pre>
[WebAuthn] Overrelease in error for multi-origin credentials
<a href="https://rdar.apple.com/123357813">rdar://123357813</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=269942">https://bugs.webkit.org/show_bug.cgi?id=269942</a>

Reviewed by Chris Dumez.

There&apos;s evidence for crashes being caused by the over-release of the dictionary inside
userInfo[NSUnderlyingErrorKey] for the error returned for multi-origin credentials. This
patch is to use RetainPtr in WebKit for this dictionary. I have not been able to reproduce
the crash, so this is speculative.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):

Canonical link: <a href="https://commits.webkit.org/275466@main">https://commits.webkit.org/275466@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79c44e90622c2598a0a901db176fa8bf49c03733

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41901 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20916 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44284 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44483 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37996 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18246 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42475 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17841 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/36085 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15527 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45886 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38085 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37434 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16716 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13730 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39668 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18335 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9391 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18394 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17979 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -369,9 +369,12 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
-                RetainPtr<NSError> underlyingError = adoptNS(error.get().userInfo[NSUnderlyingErrorKey]);
-                if ([underlyingError.get().domain isEqualToString:ASCAuthorizationErrorDomain] && underlyingError.get().code == ASCAuthorizationErrorSecurityError)
-                    exceptionData = { ExceptionCode::SecurityError, underlyingError.get().userInfo[NSLocalizedFailureReasonErrorKey] };
+                RetainPtr<NSError> underlyingError = retainPtr(error.get().userInfo[NSUnderlyingErrorKey]);
```
`<NSError>` is likely unnecessary, so is the call to `retainPtr()`

cdumez replied:
And yes, the adopting here did look like a bug to me.

cdumez replied:
`retainPtr()` is not needed.

pascoej replied:
It does not compile without it.


```
/Volumes/Data/worker/iOS-17-Simulator-Build-EWS/build/Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:401:31: error: no viable constructor or deduction guide for deduction of template arguments of 'RetainPtr'
                    RetainPtr underlyingUserInfo = underlyingError.get().userInfo;
```

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -369,9 +369,12 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
-                    exceptionData = { ExceptionCode::SecurityError, underlyingError.get().userInfo[NSLocalizedFailureReasonErrorKey] };
+                RetainPtr<NSError> underlyingError = retainPtr(error.get().userInfo[NSUnderlyingErrorKey]);
+                if ([underlyingError.get().domain isEqualToString:ASCAuthorizationErrorDomain] && underlyingError.get().code == ASCAuthorizationErrorSecurityError) {
+                    RetainPtr underlyingUserInfo = retainPtr(underlyingError.get().userInfo);
```
Unnecessary call to `retainPtr()`.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -369,9 +369,12 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
+                RetainPtr<NSError> underlyingError = retainPtr(error.get().userInfo[NSUnderlyingErrorKey]);
+                if ([underlyingError.get().domain isEqualToString:ASCAuthorizationErrorDomain] && underlyingError.get().code == ASCAuthorizationErrorSecurityError) {
+                    RetainPtr underlyingUserInfo = retainPtr(underlyingError.get().userInfo);
+                    RetainPtr failureReason = retainPtr(underlyingUserInfo.get()[NSLocalizedFailureReasonErrorKey]);
```
Unnecessary call to `retainPtr()`.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -369,9 +369,12 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
+                if ([underlyingError.get().domain isEqualToString:ASCAuthorizationErrorDomain] && underlyingError.get().code == ASCAuthorizationErrorSecurityError) {
+                    RetainPtr underlyingUserInfo = retainPtr(underlyingError.get().userInfo);
+                    RetainPtr failureReason = retainPtr(underlyingUserInfo.get()[NSLocalizedFailureReasonErrorKey]);
+                    exceptionData = { ExceptionCode::SecurityError, failureReason.get() };
```
I didn't check the type of `exceptionData` but would `WTFMove(failureReason)` reduce churn here?

pascoej replied:
The second member of `exceptionData` is a WTFString. I don't see a way in `WTFString.h` for a `RetainPtr<NSString>&&`  to work here.

cdumez reviewed (approved):
LGTM with nits.

pascoej committed:
[WebAuthn] Overrelease in error for multi-origin credentials
rdar://123357813
https://bugs.webkit.org/show_bug.cgi?id=269942

Reviewed by Chris Dumez.

There's evidence for crashes being caused by the over-release of the dictionary inside
userInfo[NSUnderlyingErrorKey] for the error returned for multi-origin credentials. This
patch is to use RetainPtr in WebKit for this dictionary. I have not been able to reproduce
the crash, so this is speculative.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):

Canonical link: https://commits.webkit.org/275466@main
